parallel-install/pkg/preinstaller/preinstaller.go

```diff
@@ -0,0 +1,127 @@
+package preinstaller
+
+import (
+	"fmt"
+	"github.com/avast/retry-go"
+	"github.com/kyma-incubator/hydroform/parallel-install/pkg/config"
+	"io/ioutil"
+	"k8s.io/client-go/dynamic"
+	"os"
+)
+
+// PreInstaller prepares k8s cluster for Kyma installation.
+type PreInstaller struct {
+	applier       ResourceApplier
+	cfg           config.Config
+	dynamicClient dynamic.Interface
+	retryOptions  []retry.Option
+}
+
+// NewPreInstaller creates a new instance of PreInstaller.
+func NewPreInstaller(applier ResourceApplier, cfg config.Config, dynamicClient dynamic.Interface, retryOptions []retry.Option) *PreInstaller {
+	return &PreInstaller{
+		applier:       applier,
+		cfg:           cfg,
+		dynamicClient: dynamicClient,
+		retryOptions:  retryOptions,
+	}
+}
+
+// InstallCRDs on a k8s cluster.
+// Returns Output containing results of installation.
+func (i *PreInstaller) InstallCRDs() (Output, error) {
+	resource := resourceType{
+		name: "crds",
+	}
+
+	output, err := i.apply(resource)
+	if err != nil {
+		return Output{}, err
+	}
+
+	return output, nil
+}
+
+// CreateNamespaces in a k8s cluster.
+// Returns Output containing results of installation.
+func (i *PreInstaller) CreateNamespaces() (Output, error) {
+	resource := resourceType{
+		name: "namespaces",
+	}
+
+	output, err := i.apply(resource)
+	if err != nil {
+		return Output{}, err
+	}
+
+	return output, nil
+}
+
+func (i *PreInstaller) apply(resourceType resourceType) (o Output, err error) {
+	installationResourcePath := i.cfg.InstallationResourcePath
+	path := fmt.Sprintf("%s/%s", installationResourcePath, resourceType.name)
+
+	rawComponentsDir, err := ioutil.ReadDir(path)
+	if err != nil {
+		return o, err
+	}
+
+	components := findOnlyDirectoriesAmong(rawComponentsDir)
```

Comment:
What for is a "component" here? I am not sure if we need any "components" in preinstall phase.

Reply:
In current approach you can find preinstaller resources in Kyma repository. They are as it follows:
```
- installation
    - resources
        - crds
            - component1
                - file 1
                - file 2
                - ...
                - file n
            - component1
                - file 1
                - file 2
                - ...
                - file n
        - namespaces
            // the same composition as for crds
```
I have introduced a mechanism for creating resources of a specific type. They should always be under appropriate subdirectory placed in `installation/resources`. However, inside these directories resources should be divided by component (for better logical division, ease of component-flagging mechanism that could be introduced etc).
So, to sum up - in this context `component` means either internal Kyma component or external Kyma dependency (which should be still so-called components).